src/servers/src/elasticsearch.rs

```diff
@@ -0,0 +1,366 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Instant;
+
+use axum::extract::{Query, State};
+use axum::headers::ContentType;
+use axum::http::{HeaderMap, HeaderName, HeaderValue, StatusCode};
+use axum::response::IntoResponse;
+use axum::{Extension, TypedHeader};
+use common_error::ext::ErrorExt;
+use common_telemetry::{debug, warn};
+use serde_json::{json, Deserializer, Value};
+use session::context::{Channel, QueryContext};
+use snafu::{ensure, ResultExt};
+
+use crate::error::{
+    status_code_to_http_status, InvalidElasticsearchInputSnafu, ParseJsonSnafu,
+    Result as ServersResult,
+};
+use crate::http::event::{
+    ingest_logs_inner, LogIngesterQueryParams, LogState, GREPTIME_INTERNAL_IDENTITY_PIPELINE_NAME,
+};
+
+// The fake version of Elasticsearch and used for `_version` API.
+const ELASTICSEARCH_VERSION: &str = "8.16.0";
+
+// Return fake response for Elasticsearch ping request.
+#[axum_macros::debug_handler]
+pub async fn handle_get_version() -> impl IntoResponse {
+    let body = serde_json::json!({
+        "version": {
+            "number": ELASTICSEARCH_VERSION
+        }
+    });
+    (StatusCode::OK, elasticsearch_headers(), axum::Json(body))
+}
+
+// Return fake response for Elasticsearch license request.
+// Reference: https://www.elastic.co/guide/en/elasticsearch/reference/current/get-license.html.
+#[axum_macros::debug_handler]
+pub async fn handle_get_license() -> impl IntoResponse {
+    let body = serde_json::json!({
+        "license": {
+            "uid": "cbff45e7-c553-41f7-ae4f-9205eabd80xx",
+            "type": "oss",
+            "status": "active",
+            "expiry_date_in_millis": 4891198687000_i64,
+        }
+    });
+    (StatusCode::OK, elasticsearch_headers(), axum::Json(body))
+}
+
+// Process `_bulk` API requests. Only support to create logs.
+// Reference: https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#docs-bulk-api-request.
+#[axum_macros::debug_handler]
+pub async fn handle_bulk_api(
+    State(log_state): State<LogState>,
+    Query(params): Query<LogIngesterQueryParams>,
+    Extension(mut query_ctx): Extension<QueryContext>,
+    TypedHeader(_content_type): TypedHeader<ContentType>,
+    payload: String,
+) -> impl IntoResponse {
+    let start = Instant::now();
+    debug!(
+        "Received bulk request, params: {:?}, payload: {:?}",
+        params, payload
+    );
+
+    // The `schema` is already set in the query_ctx in auth process.
+    query_ctx.set_channel(Channel::Elasticsearch);
+
+    // Record the ingestion time histogram.
+    let _timer = crate::metrics::METRIC_ELASTICSEARCH_LOGS_INGESTION_ELAPSED
+        .with_label_values(&[params.db.unwrap_or("public".to_string()).as_str()])
```

Comment:
Prefer `unwarp_or_else(|| "public".to_string())` to avoid `to_string` when db is present.

Reply:
TIL